packages/arb-avm-cpp/avm_values/include/avm_values/code.hpp

```diff
@@ -235,4 +349,110 @@ class Code {
     }
 };
 
+class RunningCode : public CodeBase<RunningCode>, public Code {
+    friend CodeBase<RunningCode>;
+
+    mutable std::mutex mutex;
+    uint64_t first_segment;
+    std::vector<std::shared_ptr<CodeSegment>> segment_list;
+
+    std::shared_ptr<Code> parent;
+
+    const std::shared_ptr<CodeSegment>& getSegment(uint64_t segment_num) const {
+        return segment_list.at(segment_num - first_segment);
+    }
+
+    uint64_t nextSegmentNum() const {
+        return first_segment + segment_list.size();
+    }
+
+    void storeSegment(std::shared_ptr<CodeSegment> segment) {
+        segment_list.push_back(std::move(segment));
+    }
+
+   public:
+    RunningCode(std::shared_ptr<Code> parent_)
+        : first_segment(parent_->getNextSegmentNum()),
+          parent(std::move(parent_)) {}
+
+    uint64_t fillInCode(
+        std::unordered_map<uint64_t, std::shared_ptr<CodeSegment>>&
+            parent_segments,
+        const std::map<uint64_t, uint64_t>& segment_counts) const {
+        const std::lock_guard<std::mutex> lock(mutex);
+        auto it = segment_counts.lower_bound(first_segment);
+        auto end = segment_counts.end();
+        for (; it != end; ++it) {
+            parent_segments[it->first] = getSegment(it->first);
```

Comment:
Can we assert that parent_segments doesn't have this key yet here? That would protect against collisions.